server/grpc/server.go

```diff
@@ -0,0 +1,151 @@
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+
+	commonv1pb "github.com/dapr/go-sdk/dapr/proto/common/v1"
+	pb "github.com/dapr/go-sdk/dapr/proto/runtime/v1"
+	"github.com/dapr/go-sdk/server/event"
+	"google.golang.org/grpc"
+)
+
+// NewServer creates new Server
+func NewServer(port string) (server *Server, err error) {
+	address := fmt.Sprintf(":%s", port)
+	lis, err := net.Listen("tcp", address)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to TCP listen on: %s", address)
+		return
+	}
+
+	server = &Server{
+		listener:           lis,
+		invokeHandlers:     make(map[string]func(ctx context.Context, contentTypeIn string, dataIn []byte) (contentTypeOut string, dataOut []byte)),
+		topicSubscriptions: make(map[string]func(ctx context.Context, event *event.TopicEvent) error),
+		bindingHandlers:    make(map[string]func(ctx context.Context, in *event.BindingEvent) error),
+	}
+	return
+}
+
+// Server is the gRPC server implementation for Dapr
+type Server struct {
+	listener           net.Listener
+	invokeHandlers     map[string]func(ctx context.Context, contentTypeIn string, dataIn []byte) (contentTypeOut string, dataOut []byte)
+	topicSubscriptions map[string]func(ctx context.Context, event *event.TopicEvent) error
+	bindingHandlers    map[string]func(ctx context.Context, in *event.BindingEvent) error
+}
+
+// Start registers the server and starts it
+func (s *Server) Start() error {
+	gs := grpc.NewServer()
+	pb.RegisterAppCallbackServer(gs, s)
+	return gs.Serve(s.listener)
+}
+
+// START INVOKE
+
+// Invocation
+
+// AddInvocationHandler adds provided handler to the local collection before server start
+func (s *Server) AddInvocationHandler(method string, fn func(ctx context.Context, contentTypeIn string, dataIn []byte) (contentTypeOut string, dataOut []byte)) {
+	s.invokeHandlers[method] = fn
+}
+
+// OnInvoke gets invoked when a remote service has called the app through Dapr
+func (s *Server) OnInvoke(ctx context.Context, in *commonv1pb.InvokeRequest) (*commonv1pb.InvokeResponse, error) {
+	if val, ok := s.invokeHandlers[in.Method]; ok {
+		ct, d := val(ctx, in.ContentType, in.Data.Value)
```

Comment:
What happens when caller application pass their own protobuf message object?

Reply:
This is serving side, the user specifies both the content ([]bytes] and the corresponding content type. Can you provide example when the current implementation does not work?

Reply:
content_type is valid only if data is used as bytes. for protobuf message, you need to pass in.Data which includes type_url property. 

please see this: https://github.com/dapr/dapr/blob/841849a6a15b27a1dd3877b332984d1a8ddfbbdd/dapr/proto/common/v1/common.proto#L54-L62

Reply:
ye, we can add another method to support that use-case

Reply:
Added `DataTypeURL` support to the proto content. @youngbupark PTAL

Reply:
closing since this is outdated